### PR TITLE
fix: save failed connect request urls to har files correctly

### PR DESCRIPTION
**Pull request by basedBaba**
#### Description
Fix issue #6579

This saves failed CONNECT request URLs to HAR files correctly.

#### Checklist

 - [x] I have updated tests where applicable.
 - [x] I have added an entry to the CHANGELOG.

### CODE REVIEW

**Comment by DaniElectra** [comment]
The fix works as intended :+1: 

Only thing is you have to add a test to cover the problem so that it doesn't happen again

**Comment by basedBaba** [comment]
I have added the relevant tests for the fix which required a few modifications in the sample flow generator `tflow.py`'s `twebsocketflow` function which allowed me to generate a sample flow using the `CONNECT` method and a non-existent domain for this particular case.

The checks also seem to pass now.

**Comment by basedBaba** [comment]
@DaniElectra Hello! Please have a look at this.

File: test/mitmproxy/addons/test_savehar.py
Comment (DaniElectra) [inline test/mitmproxy/addons/test_savehar.py:186]: This if statement feels redundant, as we already know the method will be CONNECT, as per the configuration above. The following should work too:

```suggestion
    assert flow_entry["request"]["url"].startswith("https")
```

**Review by DaniElectra — COMMENTED** [review]
btw, I'm not a mitmproxy maintainer but a regular user, so we'd have to wait for an actual review by one of the maintainers so that this gets merged

File: test/mitmproxy/addons/test_savehar.py
Comment (basedBaba) [reply test/mitmproxy/addons/test_savehar.py:186]: That makes sense, thanks for the suggestion!

**Comment by basedBaba** [comment]
@mhils Hello! Please have a look at this.

File: test/mitmproxy/addons/test_savehar.py
Comment (mhils) [inline test/mitmproxy/addons/test_savehar.py:179]: A WebSocket flow shouldn"t have a CONNECT method, AFAIK that's typically GET with Upgrade headers? Or is this something you've seen in the wild?

**Review by mhils — COMMENTED** [review]
Thank you two! Out of curiosity, could you explain to me how you even get CONNECT requests in mitmproxy? We typically don't surface them in the UI (or the event hooks), so I'm surprised to see them at all.

Otherwise this generally looks good, but I feel this could be a regular HTTP flow (and not a WebSocket one). In that case we don't need the changes to `tflow.py`. :)

File: test/mitmproxy/addons/test_savehar.py
Comment (DaniElectra) [reply test/mitmproxy/addons/test_savehar.py:179]: You are right, I think `ttcpflow` should be used instead? Correct me if I'm wrong

Comment (mhils) [reply test/mitmproxy/addons/test_savehar.py:179]: This should be an HTTP flow. :)

Comment (basedBaba) [reply test/mitmproxy/addons/test_savehar.py:179]: Due to my limited knowledge and understanding of the code-base, I looked at some other tests in `test_savehar.py` and concluded that `twebsocketflow` can be used for generating a sample flow without knowing it's actual use. My apologies for that.

As you said, no changes are required in `tflow.py`.

Instead, we can use `tflow` and pass a `CONNECT` request in order to generate a sample HTTP flow for testing.

File: test/mitmproxy/addons/test_savehar.py
Comment (mhils) [inline test/mitmproxy/addons/test_savehar.py:176]:
```suggestion
def test_flow_entry():
    """https://github.com/mitmproxy/mitmproxy/issues/6579"""
```

**Review by mhils — COMMENTED** [review]
Thanks! Let's add a ref for the original issue and fix the CHANGELOG conflict, then this is good to go. :)

I'd also still like to get an answer to my question in https://github.com/mitmproxy/mitmproxy/pull/6599#pullrequestreview-1830638895: How  do you get CONNECT requests visible within mitmproxy?

**Comment by mhils** [comment]
> Based on how much I understand the working of `mitmproxy`, when we send a request to a non-existent domain with the `https` protocol, `mitmproxy` first sends a `CONNECT` request to the given non-existent domain, which itself fails, resulting in the connection request being terminated, and the failed `CONNECT` request appearing on the client side. I could be completely wrong here though. :/

Can you clarify what exactly you are doing / which proxy mode you are using?

**Review by mhils — APPROVED** [review]

**Comment by basedBaba** [comment]
This is what I was trying to do.

image

image

**Comment by mhils** [comment]
Thanks for the details, that's a great repro!

**Comment by DaniElectra** [comment]
Thank you all! @mhils can you also fix the changelog and put this PR on `mitmproxy next`?

**Comment by mhils** [comment]
Good catch, thanks! 🍰

**Pull request merged**
